express-notes-api/index.js

```diff
@@ -0,0 +1,158 @@
+  noErr: 'no error'
+};
+
+function errorMsgID(id) {
+  let error;
+  let idIsThere;
+  const idNum = parseFloat(id);
+  const isInt = Number.isInteger(idNum);
+
+  if (typeof idNum === 'number' && idNum > 0 && !isNaN(idNum)) {
+    if (isInt) {
+      for (const prop in notes) {
+        if (parseInt(prop) === parseInt(id)) {
+          idIsThere = true;
+        }
+      }
+      if (idIsThere === true) {
+        error = errors.noErr;
+      } else {
+        errors.errCannotFind.error = `cannot find note with id ${id}`;
+        error = errors.errCannotFind;
+      }
+    } else {
+      error = errors.errPostiveInt;
+    }
+  } else {
+    error = errors.errPostiveInt;
+  }
+
+  return error;
+}
+
+function errorMsgContent() {
+  const error = errors.errRequireContent;
+  return error;
+}
+
+app.get('/api/notes/:id', (req, res) => {
+  const id = req.params.id;
+  const error = errorMsgID(id);
+
+  var noteToRead;
+
+  if (error === errors.errPostiveInt) {
+    res.status(400).json(error);
+  } else if (error === errors.errCannotFind) {
+    res.status(404).json(error);
+  } else if (error === errors.noErr) {
+    for (const prop in notes) {
+      if (prop === id) {
+        noteToRead = notes[prop];
+      }
+    }
+    res.status(200).json(noteToRead);
+  }
+
+});
+
+app.use(express.json());
+
+app.post('/api/notes', (req, res) => {
+  let error;
+  if (req.body.content === undefined) {
+    error = errorMsgContent();
+    res.status(400).json(error);
+  } else {
+    const newContent = req.body;
+    newContent.id = nextID;
+    notes[nextID] = newContent;
+    nextID++;
+    data.nextId = nextID;
+    data.notes = notes;
+    rewriteFile(data);
+    res.status(201).send(newContent);
+  }
+});
+
+app.delete('/api/notes/:id', (req, res) => {
+  const id = req.params.id;
+  const error = errorMsgID(id);
+
+  if (error === errors.errPostiveInt) {
+    res.status(400).json(error);
+  } else if (error === errors.errCannotFind) {
+    res.status(404).json(error);
+  } else if (error === errors.noErr) {
+    for (const prop in notes) {
+      if (prop === id) {
+        delete notes[id];
+      }
+    }
+    data.notes = notes;
+    rewriteFile(data);
+    res.status(204).send();
+  }
+});
+
+app.put('/api/notes/:id', (req, res) => {
+  const id = req.params.id;
+  let error = errorMsgID(id);
+  let newContent;
+
+  if (error === errors.errPostiveInt) {
+    res.status(400).json(error);
+  } else if (error === errors.errCannotFind) {
+    res.status(404).json(error);
+  } else if (error === errors.noErr) {
+    if (req.body.content === undefined) {
+      error = errorMsgContent();
+      res.status(400).json(error);
+    } else {
+      newContent = req.body.content;
+
+      for (const prop in notes) {
+        if (prop === id) {
+          notes[prop].content = newContent;
+          var successMsg = notes[prop];
+        }
+      }
+      data.notes = notes;
+      rewriteFile(data);
+      res.status(200).json(successMsg);
```

Comment:
It's not a good idea to send a response before the data is saved to the file (since it might error). Since `rewriteFile` is asynchronous, there's a good chance `res.json(...)` will start sending before you know if the file write was successful.